src/xmipp/applications/scripts/tomo_misalignment_resid_statistics/batch_tomo_misalignment_resid_statistics.py

```diff
@@ -0,0 +1,535 @@
+        self.residY[id].append(mData.getValue(xmippLib.MDL_SHIFT_Y, objId))
+
+      else:
+        self.residX[id] = [mData.getValue(xmippLib.MDL_SHIFT_X, objId)]
+        self.residY[id] = [mData.getValue(xmippLib.MDL_SHIFT_Y, objId)]
+
+      #Read image vectors
+      id = int(mData.getValue(xmippLib.MDL_Z, objId))
+
+      if id in self.imageX.keys():
+        self.imageX[id].append(mData.getValue(xmippLib.MDL_SHIFT_X, objId))
+        self.imageY[id].append(mData.getValue(xmippLib.MDL_SHIFT_Y, objId))
+
+      else:
+        self.imageX[id] = [mData.getValue(xmippLib.MDL_SHIFT_X, objId)]
+        self.imageY[id] = [mData.getValue(xmippLib.MDL_SHIFT_Y, objId)]
+      
+    # Debug
+    if self.checkParam('--debug'):
+      self.debug=True
+
+    print("Residual information read successfully!")
+               
+
+  def writeOutputStatsInfo(self, residualStats, filePath):
+    """
+      Write statistical information into metadata
+    """
+
+    print("Writting output statistics at " + filePath)
+
+    mData = xmippLib.MetaData()
+
+    for i in range(len(residualStats)):
+      id = mData.addObject()
+      mData.setValue(xmippLib.MDL_ENABLED, residualStats[i][0], id)
+      mData.setValue(xmippLib.MDL_MIN, residualStats[i][1], id)
+      mData.setValue(xmippLib.MDL_MAX, residualStats[i][2], id)
+      mData.setValue(xmippLib.MDL_IMAGE, residualStats[i][3], id)
+
+    mData.write(filePath)
+
+
+  def generateSideInfo(self):
+    """
+      Generate residual side information to perform posterior tests
+    """
+
+    # Residual info vectors for landmarks
+    if self.debug:
+      print("Generate side information for landmarks...")
+
+    for key in self.residX.keys():
+      self.residSize[key] = len(self.residX[key])
+
+      nPosX = 0
+      nPosY = 0
+
+      for i, r in enumerate(self.residX[key]):
+        if r > 0:
+          nPosX += 1
+
+        if i == 0:
+          self.residXAcc[key] = [r]
+        else:
+          self.residXAcc[key].append(r+self.residXAcc[key][i-1])
+
+      for i, r in enumerate(self.residY[key]):
+        if r > 0:
+          nPosY += 1
+
+        if i == 0:
+          self.residYAcc[key] = [r]
+        else:
+          self.residYAcc[key].append(r+self.residYAcc[key][i-1])
+
+      self.nResidPosX[key] = nPosX
+      self.nResidPosY[key] = nPosY
+
+      for i in range(self.residSize[key]):
+        if i == 0:
+          self.residModuleAcc[key] = [sqrt(self.residXAcc[key][i]*self.residXAcc[key][i] + self.residYAcc[key][i]*self.residYAcc[key][i])]
+        else:
+          self.residModuleAcc[key].append(sqrt(self.residXAcc[key][i]*self.residXAcc[key][i] + self.residYAcc[key][i]*self.residYAcc[key][i]))
+
+    # Residual info vectors for images
+    if self.debug:
+      print("Generate side information for images...")
+
+    for key in self.imageX.keys():
+      self.imageSize[key] = len(self.imageX[key])
+
+      nPosX = 0
+      nPosY = 0
+
+      for i, r in enumerate(self.imageX[key]):
```

Comment:
This function is repeated in the code, perhaps it can be... nPosY += len(self.imageX[key]>0) avoiding the for loop?

Reply:
I do agree that this removes the nPosY += 1 from the loop but hese operations:

        if i == 0:
          self.residXAcc[key] = [r]
        else:
          self.residXAcc[key].append(r+self.residXAcc[key][i-1])
          
are still needed and nested in the loop, to it wouldn't be removed. Also, the execution of this script is very fast so the increment in performance in neglectable.